### PR TITLE
Apply portal host into data plane

### PR DESCRIPTION
**Pull request by jshaofuturewei**
The PR is to add portal host ip to data plane. There are some changes in control plane which are pass portal host ip to rpc. The change is saved in subnet map and can be picked up by xdp.

Create slave.in and portal_host.properties
```bash
root@ip-172-31-2-32:~/mizar_cluster_scripts# cat slaves.in
52.43.138.8
54.189.90.175
52.12.138.187
root@ip-172-31-2-32:~/mizar_cluster_scripts# cat portal_host.properties
172.31.15.208
```
Update vanilla-net1.yaml and vanilla-net2.yaml to 2 slave hosts
```bash
root@ip-172-31-2-32:~/mizar_cluster_scripts# cat vanilla-net1.yaml
apiVersion: v1
kind: Pod
metadata:
  name: pod-in-net1
  annotations:
    mizar.com/vpc: vpc1
    mizar.com/subnet: net1
spec:
  containers:
  - name: vanilla-container
    image: mizarnet/testpod:latest
  nodeName: ip-172-31-15-85
root@ip-172-31-2-32:~/mizar_cluster_scripts# cat vanilla-net2.yaml
apiVersion: v1
kind: Pod
metadata:
  name: pod-in-net2
  annotations:
    mizar.com/vpc: vpc1
    mizar.com/subnet: net2
spec:
  containers:
  - name: vanilla-container
    image: mizarnet/testpod:latest
  nodeName: ip-172-31-4-20
```
Set up the cluster
```bash
./build_docker_image.sh && ./restart_cluster.sh && ./create_vpc_nets.sh && ./create_pods.sh
```
Check the output and find that portal_host is added to configmap
```bash
Replacing [PORTAL_HOST] with 172.31.15.208 in configmap
configmap/portal-host-config created
```
Verify the change using kubectl
```bash
root@ip-172-31-2-32:~/mizar_cluster_scripts# kubectl get configmap portal-host-config -o json
{
    "apiVersion": "v1",
    "data": {
        "portal_host_ip": "172.31.15.208"
    },
    "kind": "ConfigMap",
    "metadata": {
        "annotations": {
            "kubectl.kubernetes.io/last-applied-configuration": "{\"apiVersion\":\"v1\",\"data\":{\"portal_host_ip\":\"172.31.15.208\"},\"kind\":\"ConfigMap\",\"metadata\":{\"annotations\":{},\"name\":\"portal-host-config\",\"namespace\":\"default\"}}\n"
        },
        "creationTimestamp": "2021-11-15T19:03:37Z",
        "name": "portal-host-config",
        "namespace": "default",
        "resourceVersion": "1197",
        "uid": "0f8b970a-642a-402c-8052-3a4310928669"
    }
}
```
Run the follow command to get nodes
```bash
root@ip-172-31-2-32:~/mizar_cluster_scripts# kubectl get nodes
```
And you will get the result like
```bash
NAME               STATUS   ROLES                  AGE   VERSION
ip-172-31-15-208   Ready    <none>                 32m   v1.22.2
ip-172-31-15-85    Ready    <none>                 32m   v1.22.2
ip-172-31-2-32     Ready    control-plane,master   33m   v1.22.2
ip-172-31-4-20     Ready    <none>                 32m   v1.22.2

```
Run the following commands to get current bouncers and dividers
```bash
kubectl get bouncers
kubectl get dividers
```
And you will get the result with 3 bouncers and 2 dividers.  net2 is a subnet which is external so that its bouncer has been set to  the portal host 172.31.15.208  we defined at beginning.
```bash
net0-b-43368556-1a0d-492b-b6d4-f3acc6cc2df6   vpc0   net0   172.31.15.85    0a:a7:d2:0e:4b:bb   ip-172-31-15-85    Provisioned   2021-11-15T19:03:36.778236   1.679017
net1-b-5a5ba990-d6d9-4774-9cdf-68c63ab9fda8   vpc1   net1   172.31.15.85    0a:a7:d2:0e:4b:bb   ip-172-31-15-85    Provisioned   2021-11-15T19:03:39.305300   0.786885
net2-b-65aabc42-4132-4434-a05d-66a40513401e   vpc1   net2   172.31.15.208   0a:c5:74:b5:4b:65   ip-172-31-15-208   Provisioned   2021-11-15T19:03:39.512262   0.914433
```
```bash
NAME                                          VPC    IP             MAC                 DROPLET           STATUS        CREATETIME                   PROVISIONDELAY
vpc0-d-318efaca-69f3-427d-8bb9-52f3a04304e4   vpc0   172.31.15.85   0a:a7:d2:0e:4b:bb   ip-172-31-15-85   Provisioned   2021-11-15T19:03:16.662336   0.307799
vpc1-d-5d56a625-4672-403b-876a-f94e33bcd852   vpc1   172.31.15.85   0a:a7:d2:0e:4b:bb   ip-172-31-15-85   Provisioned   2021-11-15T19:03:38.082251   0.252435
```
It can be verified by checking control plane logs
```bash
root@ip-172-31-2-32:~/mizar_cluster_scripts# kubectl logs mizar-operator-5b4fd665f5-d8qgc | grep "portal"
[2021-11-16 17:28:48,846] root                 [INFO    ] The portal host ip is 172.31.15.208
[2021-11-16 17:28:48,846] root                 [INFO    ] A droplet 172.31.15.208 has been added as portal.
[2021-11-16 17:28:48,846] root                 [INFO    ] The portal droplet <mizar.obj.droplet.Droplet object at 0x7f302ef1d390> has been removed.
[2021-11-16 17:28:48,846] root                 [INFO    ] external subnet, using portal droplet 172.31.15.208
[2021-11-16 17:28:48,897] root                 [INFO    ] The portal host ip is 172.31.15.208
[2021-11-16 17:28:48,921] root                 [INFO    ] update_net: /trn_bin/transit -s 172.31.4.20  update-net -i eth0 -j '{"tunnel_id": "2", "nip": "192.168.122.0", "prefixlen": "24", "switches_ips": ["172.31.15.208"], "external": true, "portal_host": "172.31.15.208"}'
```
Verify the portal host ip config is passed to xdp
```bash
root@ip-172-31-2-32:~/mizar_cluster_scripts# cat /var/log/syslog | grep "portal"
Nov 16 17:28:45 ip-172-31-2-32 transit[3666]: [update_net_1_svc:166] update_net_1 net tunid: 1, netip: 0x14, prefixlen: 8, with 1 switches, portal 0x14041fac
```
Ping each other

```bash
root@ip-172-31-2-32:~/mizar_cluster_scripts# ./ping_0.sh
-------
pinging 192.168.0.2 from 192.168.122.6
-------
PING 192.168.0.2 (192.168.0.2) 56(84) bytes of data.
64 bytes from 192.168.0.2: icmp_seq=1 ttl=64 time=1.13 ms
64 bytes from 192.168.0.2: icmp_seq=2 ttl=64 time=0.590 ms


root@ip-172-31-2-32:~/mizar_cluster_scripts# ./ping_122.sh
-------
pinging 192.168.122.6 from 192.168.0.2
-------
PING 192.168.122.6 (192.168.122.6) 56(84) bytes of data.
64 bytes from 192.168.122.6: icmp_seq=1 ttl=64 time=0.490 ms
64 bytes from 192.168.122.6: icmp_seq=2 ttl=64 time=0.446 ms
```

### CODE REVIEW

**Comment by pdgetrf** [comment]
looks good over all. 

two small suggestion: 

1. in the description, add a small paragraph at the beginning giving a brief explanation what this PR does, and then continue with the rest of the steps which is to show the validation steps
2. when and how is the config map portal-host-config created? the description only mentions checking it.

File: src/rpcgen/trn_rpc_protocol.x [hunk 1]
```diff
@@ -50,6 +50,7 @@ struct rpc_trn_network_t {
        uint64_t tunid;
        uint32_t netip;
        uint32_t switches_ips<RPC_TRN_MAX_NET_SWITCHES>;
+       uint32_t portal_host;
```
Comment (pdgetrf) [inline src/rpcgen/trn_rpc_protocol.x]: nit: portal_host --> portal_host_ip

**Review by pdgetrf — APPROVED** [review]
minor suggestion for better code readability and two questions in the comments.

File: src/rpcgen/trn_rpc_protocol.x (hunk 1 above)
Comment (jshaofuturewei) [reply src/rpcgen/trn_rpc_protocol.x]: Updated

**Comment by jshaofuturewei** [comment]
> looks good over all.
> 
> two small suggestion:
> 
> 1. in the description, add a small paragraph at the beginning giving a brief explanation what this PR does, and then continue with the rest of the steps which is to show the validation steps
> 2. when and how is the config map portal-host-config created? the description only mentions checking it.

Updated

**Pull request merged**